### PR TITLE
[Quantization] Quanto quantizer

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do ? 
This PR adds the quantization methods from quanto library. We will support inference + model quantization if the user perform weights only quantization since we don't require a calibration dataset. 

TODO:
- [x] docs 
- [x] tests
- [x] Guard against saving now since it is a bit complicated

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Clean work already ! Looking forward to merge the PR !

**Comment by younesbelkada** [comment]
cc @dacorvo

File: src/transformers/quantizers/quantizer_quanto.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+class QuantoHfQuantizer(HfQuantizer):
+    """
+    Quantizer for the quanto library
+    """
+
+    required_packages = ["quanto", "accelerate"]
+
+    def __init__(self, quantization_config: QuantoConfig, **kwargs):
+        self.requires_calibration = quantization_config.activations is not None
```
[the remark this answers is not in the document]
Comment (SunMarc) [reply src/transformers/quantizers/quantizer_quanto.py:41]: For the transformers integration, we will only support quantization for weight only quantization since : 
- it does not require calibration data.
- Don't need to retrain most of the time. 
- We can perform the weight only quantization on the spot (just like bnb by quantizing each layer as we load them) -> memory efficient. 

However, for the activation, we need the whole model to be loaded and pass the calibration data. Hence, there is no real benefit of supporting the quantization of the activation in transformers. I think that it will be easier for the user to directly use quanto + perform QAT if needed. LMK if it makes sense ! If the user requests it, we can still add it. 
cc @younesbelkada

File: src/transformers/quantizers/quantizer_quanto.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    def _process_model_after_weight_loading(self, model):
+        if not self.pre_quantized:
+            from quanto import freeze, quantize
+
+            quantize(model, weights=self.quantization_config.weights, activations=self.quantization_config)
+            freeze(model)
+
+    @property
+    def is_trainable(self, model: Optional["PreTrainedModel"] = None):
+        return False
```
[the remark this answers is not in the document]
Comment (SunMarc) [reply src/transformers/quantizers/quantizer_quanto.py:82]: > If you don't freeze the weights immediately after export, then the model is trainable. But that might not be your primary goal in that first integration (since you would need to extend the HfQuantizer API to add a freeze method or take a decision when to freeze the weights).

Yes, for this integration, we will freeze the weights. I think that for more complex use cases (QAT, activation quantization), we should let the user use quanto directly. It will be more intuitive than passing args to the Quantizer. We will detail everything on a blogpost !

File: src/transformers/quantizers/quantizer_quanto.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/quantizers/quantizer_quanto.py:41]: yes, I think we can discuss about it but currently I think in terms of UX we should probably not silently calibrate the model by default, I agree we should do it only when users expliclty asks for it, what do you think @dacorvo ?
I think the main caveat of not calibrating the model under the hood is the amount of time it might take for large models to just load the model

File: src/transformers/utils/quantization_config.py [hunk 3]
```diff
@@ -792,3 +793,35 @@ def post_init(self):
...
+    """
+
+    def __init__(
+        self,
+        weights=torch.int8,
+        activations=None,
+        **kwargs,
+    ):
+        self.quant_method = QuantizationMethod.QUANTO
+        # TODO: I think that we can keep the naming but it will be easier for the user to set str and not torch.int8 or quanto.int4 for ex
```
[the remark this answers is not in the document]
Comment (dacorvo) [reply src/transformers/utils/quantization_config.py]: It is done. I have also another pull-request corresponding to the issue you found, but with a different fix.

**Comment by SunMarc** [comment]
Quick update, all tests are passing with the exception of safetensors tests. I've also implemented the quantization on the spot just like we do for bnb quantization. However, if the user wants to use cpu/disk offload, he will need to install the main branch of accelerate for now because of this [PR](https://github.com/huggingface/accelerate/pull/2481).

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -367,6 +367,10 @@ def shard_checkpoint(
         # check: https://github.com/huggingface/transformers/pull/24416 for more details
         if isinstance(weight, str):
             continue
+        elif weight.__class__.__name__ in ["QTensor"]:
```
Comment (younesbelkada) [inline src/transformers/modeling_utils.py:370]:
```suggestion
        elif weight.__class__.__name__ in ["QTensor", "QBitsTensor"]:
```

File: src/transformers/quantizers/quantizer_quanto.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            if torch.cuda.is_available():
+                torch_dtype = torch.float16
+                logger.info(
+                    "CUDA available. Assuming Quanto inference on GPU and loading the model in `torch.float16`. To overwrite it, set `torch_dtype` manually."
+                )
+            else:
+                torch_dtype = torch.float32
+                logger.info(
+                    "CUDA is unavailable. Assuming AQLM inference on CPU and loading the model in `torch.float32`. To overwrite it, set `torch_dtype` manually."
+                )
```
Comment (younesbelkada) [inline src/transformers/quantizers/quantizer_quanto.py:85]:
```suggestion
```
I suggest we don't do anything here because quanto is quite agnostic to CPU / GPU devices, what do you think?

File: docs/source/en/quantization.md [hunk 6]
```diff
@@ -26,6 +26,61 @@ Interested in adding a new quantization method to Transformers? Read the [HfQuan
 
 </Tip>
 
+## Quanto
+
+[🤗 Quanto](https://github.com/huggingface/quanto) library is a versatile pytorch quantization toolkit. The quantization method used is the linear quantization. Quanto provides several unique features such as:
```
Comment (younesbelkada) [inline docs/source/en/quantization.md:31]: TODO (in the future) add the link to the colab notebook + mention that quanto is modality agnostic (most users will tend to think that this is only for LLMs)

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -2255,7 +2265,9 @@ def save_pretrained(
 
         hf_quantizer = getattr(self, "hf_quantizer", None)
         quantization_serializable = (
-            hf_quantizer is not None and isinstance(hf_quantizer, HfQuantizer) and hf_quantizer.is_serializable
+            hf_quantizer is not None
+            and isinstance(hf_quantizer, HfQuantizer)
+            and hf_quantizer.is_serializable(safe_serialization=safe_serialization)
```
Comment (younesbelkada) [inline src/transformers/modeling_utils.py:2270]: hmmm even though the HfQuantizer classes are not public maybe we should stay on the safe zone and not modify the propery `is_serializable` and keep it as a property. What about creating a new property `is_safe_serializable`, that by default points to `is_serializable` and we overwrite it only for quanto quantizer?

File: src/transformers/quantizers/base.py [hunk 8]
```diff
@@ -186,9 +208,8 @@ def _process_model_before_weight_loading(self, model, **kwargs):
     def _process_model_after_weight_loading(self, model, **kwargs):
         ...
 
-    @property
     @abstractmethod
-    def is_serializable(self):
+    def is_serializable(self, **kwargs):
```
Comment (younesbelkada) [inline src/transformers/quantizers/base.py:212]: See my comment above 🙏

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this huge work ! I left some comments about the property method being removed (even though it is not really breaking) + removing the logic in `update_torch_dtype`. Apart from that the PR looks in a great state IMO ! What do you think?

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (younesbelkada) [reply src/transformers/modeling_utils.py:370]: Otherwise serializing 4-bit weights fails

Comment (SunMarc) [reply src/transformers/modeling_utils.py:370]: Thanks for the catch ! I will add tests for 4 bit

File: src/transformers/modeling_utils.py (hunk 7 above)
Comment (SunMarc) [reply src/transformers/modeling_utils.py:2270]: Sounds good ! I'll revert the changes I did for the other quantizer and create a new property as you suggested !

File: src/transformers/modeling_utils.py [hunk 9]
```diff
@@ -376,7 +380,11 @@ def shard_checkpoint(
             sharded_state_dicts[block_id][key] = weight
             continue
 
-        weight_size = weight.numel() * dtype_byte_size(weight.dtype)
+        if weight.__class__.__name__ in ["QTensor", "QBitsTensor"]:
+            # we don't add the scales/zero-point since these are negligeable
```
Comment (dacorvo) [inline src/transformers/modeling_utils.py:384]: Not so much anymore with group-wise quantization, so you might consider adding them after all (and we might want to pack zeropoint in quanto too since it is now bigger).

File: src/transformers/quantizers/quantizer_quanto.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+            self.modules_to_not_convert = get_keys_to_not_convert(model)
+        else:
+            self.modules_to_not_convert = self.quantization_config.modules_to_not_convert
+
+        if not isinstance(self.modules_to_not_convert, list):
+            self.modules_to_not_convert = [self.modules_to_not_convert]
+
+        self.modules_to_not_convert.extend(keep_in_fp32_modules)
+
+        model, _ = replace_with_quanto_layers(
```
Comment (dacorvo) [inline src/transformers/quantizers/quantizer_quanto.py:178]: Just curious: why did you need to create that additional method ? Is it just because `quanto.quantize` takes a white-list of target modules to quantize and you prefer to use a black-list to exclude modules ?

File: src/transformers/utils/quantization_config.py [hunk 11]
```diff
@@ -789,3 +790,44 @@ def post_init(self):
...
+        self.activations = activations
+        self.modules_to_not_convert = modules_to_not_convert
+        self.post_init()
+
+    def post_init(self):
+        r"""
+        Safety checker that arguments are correct
+        """
+        accepted_weights = ["float8", "int8", "int4", "int2"]
+        accepted_activations = [None, "int8", "float8_e4m3", "float8_e5m2"]
```
Comment (dacorvo) [inline src/transformers/utils/quantization_config.py:829]: As said before, accept also "float8", and maybe just drop the others.

File: tests/quantization/quanto_integration/test_quanto.py [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+                nb_qlinear += 1
+
+        self.assertEqual(self.nb_linear, nb_qlinear)
+
+    def test_weight_and_activation_quantization_conversion(self):
+        """
+        Simple test that checks if the quantized model has been converted properly when using weight + activation quantization
+        """
+
+        # Try with weight + activatioin quantization
```
Comment (dacorvo) [inline tests/quantization/quanto_integration/test_quanto.py:79]:
```suggestion
        # Try with weight + activation quantization
```

File: tests/quantization/quanto_integration/test_quanto.py [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+        self.model, _ = replace_with_quanto_layers(self.model, quantization_config=quantization_config)
+
+        nb_qlinear = 0
+        for module in self.model.modules():
+            if isinstance(module, QLinear):
+                nb_qlinear += 1
+
+        self.assertEqual(self.nb_linear, nb_qlinear)
+
+    def test_weight_and_activation_quantization_conversion(self):
```
Comment (dacorvo) [inline tests/quantization/quanto_integration/test_quanto.py:74]: This is supposed to be used only when the model has been exported through quanto directly, right ?

File: tests/quantization/quanto_integration/test_quanto.py [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+        self.assertEqual(self.nb_linear - 1, nb_qlinear)
+
+
+@slow
+@require_torch_gpu
+@require_quanto
+@require_accelerate
+class QuantoQuantizationTest(unittest.TestCase):
+    """
+    Test 8-bit weights only qunatization
```
Comment (dacorvo) [inline tests/quantization/quanto_integration/test_quanto.py:122]:
```suggestion
    Test 8-bit weights only quantization
```

File: tests/quantization/quanto_integration/test_quanto.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+        from quanto import freeze, qint4, qint8, quantize
+
+        w_mapping = {"int8": qint8, "int4": qint4}
+        model = AutoModelForCausalLM.from_pretrained(
+            self.model_name,
+            device_map=self.device_map,
+            torch_dtype=torch.float32,
+        )
+        # we do not quantize the lm_head since we don't do that in transformers
+        quantize(model.transformer, weights=w_mapping[self.weights])
```
Comment (dacorvo) [inline tests/quantization/quanto_integration/test_quanto.py:258]: Yup, this is the way to do it. But I agree this is model/architecture specific.

**Review by dacorvo — COMMENTED** [review]
Nice ! A few nits and questions.

File: docs/source/en/quantization.md [hunk 16]
```diff
@@ -26,6 +26,61 @@ Interested in adding a new quantization method to Transformers? Read the [HfQuan
...
+- weights quantization (`float8`,`int8`,`int4`,`int2`)
+- activation quantization (`int8`,`float8_e4m3`,`float8_e5m2`)
```
[the remark this answers is not in the document]
Comment (SunMarc) [reply docs/source/en/quantization.md:33]: Sounds good ! I'll remove `float8_e4m3` and `float8_e5m2` from the activation list

File: src/transformers/modeling_utils.py (hunk 9 above)
Comment (SunMarc) [reply src/transformers/modeling_utils.py:384]: Let's not add too much code related to quanto, especially if we decide to pack the zero-point or quantize the scale again in the future. I think that we might be off by one GB per file (shard_size of 5GB), so that should be fine. LMK if this sounds reasonable.

File: src/transformers/utils/quantization_config.py (hunk 11 above)
Comment (SunMarc) [reply src/transformers/utils/quantization_config.py]: fixed !

File: tests/quantization/quanto_integration/test_quanto.py (hunk 13 above)
Comment (SunMarc) [reply tests/quantization/quanto_integration/test_quanto.py:74]: Yes, that's right.

**Review by dacorvo — APPROVED** [review]
LGTM, Thank you very much for that pull-request. Look forward to see new quanto weights for models on the hub !

File: src/transformers/quantizers/quantizer_quanto.py (hunk 10 above)
Comment (younesbelkada) [reply src/transformers/quantizers/quantizer_quanto.py:175]: I don't mind changing either! does `quanto.quantize` supports specifying module names to not convert?

File: docs/source/en/quantization.md [hunk 17]
```diff
@@ -26,6 +26,61 @@ Interested in adding a new quantization method to Transformers? Read the [HfQuan
...
+The integration with transformers only supports weights quantization. For the more complex use case such as activation quantization, calibration and quantization aware training, you should use quanto library instead. 
+<!-- -> add link to quanto library.  -->
+
+```py
+from transformers import AutoModelForCausalLM, AutoTokenizer, QuantoConfig
+
+model_id = "facebook/opt-125m"
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+quanto_config = QuantoConfig(weights="int8")
+quantized_model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", quantization_config=quanto_config)
```
Comment (ArthurZucker) [inline docs/source/en/quantization.md:60]:
```suggestion
quantization_config = QuantoConfig(weights="int8")
quantized_model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", quantization_config= quantization_config)
```

File: docs/source/en/quantization.md [hunk 18]
```diff
@@ -26,6 +26,61 @@ Interested in adding a new quantization method to Transformers? Read the [HfQuan
...
+```
+
+You can also save the quantized model using the [`~PreTrainedModel.save_pretrained`] method. 
+
+```py
+from transformers import AutoModelForCausalLM, AutoTokenizer, QuantoConfig
+
+quantized_model.save_pretrained(quantized_model, safe_serialization=False)
```
Comment (ArthurZucker) [inline docs/source/en/quantization.md:68]: can we disable safetensors saving in the `quanto.py`?

File: src/transformers/modeling_utils.py [hunk 19]
```diff
@@ -377,7 +381,11 @@ def shard_checkpoint(
             sharded_state_dicts[block_id][key] = weight
             continue
 
-        weight_size = weight.numel() * dtype_byte_size(weight.dtype)
+        if weight.__class__.__name__ in ["QTensor", "QBitsTensor"]:
+            # we don't add the scales/zero-point because those can be negligeable + subject to changes
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:385]: This means we are stuck with this if else for FC? if things are subject to change I don't really think we should have a if else here!

File: src/transformers/modeling_utils.py [hunk 20]
```diff
@@ -497,7 +505,7 @@ def load_sharded_checkpoint(model, folder, strict=True, prefer_safe=True):
     return torch.nn.modules.module._IncompatibleKeys(missing_keys, unexpected_keys)
 
 
-def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
+def load_state_dict(checkpoint_file: Union[str, os.PathLike], hf_quantizer=None):
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:508]: that is re-introducing quantizer related logic in the modeling utils, can't we pass `weights_only_kwarg` as arg here?

File: src/transformers/quantizers/base.py [hunk 21]
```diff
@@ -99,6 +110,16 @@ def adjust_target_dtype(self, torch_dtype: "torch.dtype") -> "torch.dtype":
...
 
+    def update_missing_keys(self, model, missing_keys: List[str], prefix: str) -> List[str]:
+        """
+        Override this method if you want to adjust the `missing_keys`.
+
+        Args:
+            missing_keys (`List[str]`, *optional*):
+                The list of missing keys in the checkpoint compared to the state dict of the model
+        """
+        return missing_keys
```
Comment (ArthurZucker) [inline src/transformers/quantizers/base.py:121]: Could you explain why we need this, only quanto requires it vs the other schemes that do not -> quanto is not consistent with other saving formats? (naive q on my side!)

File: src/transformers/modeling_utils.py [hunk 22]
```diff
@@ -3737,6 +3749,9 @@ def _fix_key(key):
             for pat in cls._keys_to_ignore_on_load_unexpected:
                 unexpected_keys = [k for k in unexpected_keys if re.search(pat, k) is None]
 
+        if hf_quantizer is not None:
+            missing_keys = hf_quantizer.update_missing_keys(model, missing_keys, prefix)
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:3753]: Not sure this is something we want / I understand why this is needed

File: src/transformers/quantizers/quantizer_quanto.py [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+                        "You have a version of `accelerate` that is not compatible cpu/disk offload with quanto quantized model. "
+                        "You need to install a version of accelerate > 0.27.0."
+                    )
+
+    def update_torch_dtype(self, torch_dtype: "torch.dtype") -> "torch.dtype":
+        return torch_dtype
+
+    def update_weights_only_kwarg(self, weights_only_kwarg: Dict[str, Any]) -> Dict[str, Any]:
+        weights_only_kwarg["weights_only"] = False
+        return weights_only_kwarg
```
Comment (ArthurZucker) [inline src/transformers/quantizers/quantizer_quanto.py:76]: let's document why we can't just load the weights ("because quanto relies on special structures") and also if there is .bin then we have safety dependency with quanto

File: tests/quantization/quanto_integration/test_quanto.py [hunk 24]
```diff
@@ -0,0 +1,426 @@
...
+            quantized_model_from_saved = AutoModelForCausalLM.from_pretrained(
+                tmpdirname,
+                device_map=self.device_map,
+                torch_dtype=torch.float32,
+            )
+        self.check_same_model(model, quantized_model_from_saved)
+        self.check_inference_correctness(quantized_model_from_saved, device="cuda")
+
+
+class QuantoQuantizationOffloadTest(QuantoQuantizationTest):
```
Comment (ArthurZucker) [inline tests/quantization/quanto_integration/test_quanto.py:292]: does this test require torch_gpu?

**Comment by ArthurZucker** [comment]
LGTM, mostly concerned with quanto specific function addition, that should leave in quanto.py not in the modeling (as much as possible of course).

File: src/transformers/quantizers/quantizer_quanto.py (hunk 23 above)
Comment (dacorvo) [reply src/transformers/quantizers/quantizer_quanto.py:76]: @SunMarc correct me if I am wrong but I think that is because when activations are quantized, quanto adds two buffers to quantized modules to store the input and output quantization scales (for weights also, but these are stored directly __inside__ the Tensor subclass with the same name, so it is transparent).
The serialized dict therefore contains more names.

File: docs/source/en/quantization.md (hunk 18 above)
Comment (SunMarc) [reply docs/source/en/quantization.md:68]: Yes, it is disabled using the `hf_quantizer.is_safe_serializable`. We will raise an error if they try to safe using safetensors !

File: src/transformers/modeling_utils.py (hunk 19 above)
Comment (SunMarc) [reply src/transformers/modeling_utils.py:385]: > if things are subject to change I don't really think we should have a if else here!

Oh, it mean that it can change for the _scales and the _zeropoint, so I decided to not add them. For _data, it shouldn't change. I will try to find a better solution. This is not pretty indeed.

File: src/transformers/quantizers/base.py (hunk 21 above)
Comment (SunMarc) [reply src/transformers/quantizers/base.py:121]: This is linked to @dacorvo answer below : 
> @SunMarc correct me if I am wrong but I think that is because when activations are quantized, quanto adds two buffers to quantized modules to store the input and output quantization scales (for weights also, but these are stored directly inside the Tensor subclass with the same name, so it is transparent).

So, after converting replacing the layers with the quantized layers, we have two additional buffers for each weight that are missing from in the checkpoint.

File: src/transformers/modeling_utils.py (hunk 22 above)
Comment (SunMarc) [reply src/transformers/modeling_utils.py:3753]: I can try to sneak that in inside the `hf_quantizer.create_quantized_param` just like how bnb deal with `unexpected_keys`.

File: src/transformers/quantizers/quantizer_quanto.py (hunk 23 above)
Comment (SunMarc) [reply src/transformers/quantizers/quantizer_quanto.py:76]: This is an issue compatibility with the safe serialization used by pytoch due to quanto special structure. You can check the issue I wrote [here](https://github.com/huggingface/quanto/issues/93) for more information ! 

@dacorvo, what you described is linked to the issue with the missing keys. I've added that comment to the related section. Thanks for explaining !

File: tests/quantization/quanto_integration/test_quanto.py (hunk 24 above)
Comment (SunMarc) [reply tests/quantization/quanto_integration/test_quanto.py:292]: Yes, but since it is a subclass of QuantoQuantizationTest and this class have the @require_torch_gpu decorator, that should be fine no ?

File: src/transformers/modeling_utils.py (hunk 20 above)
Comment (SunMarc) [reply src/transformers/modeling_utils.py:508]: I'll fix this

Comment (SunMarc) [reply src/transformers/modeling_utils.py]: Let me know if you prefer with these [changes](https://github.com/huggingface/transformers/pull/29023/commits/5fc659ce0c5008a842fe6821e30fd7b0e0f8eedd) ! Otherwise I can revert it.

File: src/transformers/modeling_utils.py (hunk 19 above)
Comment (SunMarc) [reply src/transformers/modeling_utils.py]: Let me know if this [solution](https://github.com/huggingface/transformers/pull/29023/commits/15f7a2a9b9a13d0efd9736b0186994c286f73490) is enough. We shouldn't have any FC issues.

File: src/transformers/modeling_utils.py (hunk 22 above)
Comment (SunMarc) [reply src/transformers/modeling_utils.py:3762]: The buffer are created at the moment we replace the layers so modifying the missing keys here `hf_quantizer.create_quantized_param` don't make sense + it does not cover some edges cases . If possible, I prefer leaving it like this. LMK if this is possible or not.

Comment (ArthurZucker) [reply src/transformers/modeling_utils.py:3731]: Alright no worries 🤗

File: tests/quantization/quanto_integration/test_quanto.py [hunk 25]
```diff
@@ -0,0 +1,426 @@
+# coding=utf-8
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (ArthurZucker) [inline tests/quantization/quanto_integration/test_quanto.py:2]:
```suggestion
# Copyright 2024 The HuggingFace Team. All rights reserved.
```

**Review by ArthurZucker — APPROVED** [review]
Thanks for iterating! 🔥 let's make quanto go brrr

**Review by younesbelkada — APPROVED** [review]
Thanks again @SunMarc - great work ! 🚀 let's 🚢 it

**Comment by SunMarc** [comment]
Note to the core mainteners and especially @ArthurZucker . I reverted a lot of changes I did for the serialization. I decided to postpone the serialization feature by setting the `is_serializable` property to `False` in this PR. 

I will do a follow up PR for the serialization since the [PR](https://github.com/huggingface/quanto/pull/116) to make it compatible with safetensors or `weight_only` saving in quanto needs important changes in transformers. 

Without serialization, this PR should still be good to merge !
